# Week-1 Assignment

*Welcome to your first assignment for the SimuTech Winter Project 2022! I hope you are excited to implement and test everything you have learned up until now. There is an interesting set of questions for you to refine your acquired skills as you delve into hands-on coding and deepen your understanding of numpy, pandas, and data visualization libraries.*

# Section 0 : Importing Libraries

*Let's begin by importing numpy, pandas and matplotlib.*

In [55]:
import numpy as np
import pandas as pd

# Section 1 : Playing with Python and Numpy

### Q1. Matrix Multiplication

##### (i) Check if matrix multiplication is valid

In [1]:
def isValid(A,B):
  num_columns_A = len(A[0]) 
  num_rows_B = len(B)
  return num_columns_A==num_rows_B


##### (ii) Using loops (without using numpy)

In [5]:
def matrix_multiply(A,B):
  product1= [[0 for p in range(len(B[0]))] for q in range(len(A))]

  for i in range(len(A)):
    for j in range(len(B[0])):
      for k in range(len(B)):
        product1[i][j] += A[i][k] * B[k][j]
  
  return product1



##### (iii) Using numpy

In [7]:
def matrix_multiply_2(A,B):
  product2 = np.matmul(A,B)
  return product2

##### (iv) Testing your code

Run the following cell to check if your functions are working properly.

*Expected output:*
[ [102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600] ]

In [8]:
A = np.array([
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9],
    [10, 11, 12]
])

B = np.array([
    [13, 14, 15],
    [16, 17, 18],
    [19, 20, 21]
])

if isValid(A,B):
  print(f"Result using loops: {matrix_multiply(A,B)}")
  print(f"Result using numpy: {matrix_multiply_2(A,B)}")
else:
  print(f"Matrix multiplication is not valid")

Result using loops: [[102, 108, 114], [246, 261, 276], [390, 414, 438], [534, 567, 600]]
Result using numpy: [[102 108 114]
 [246 261 276]
 [390 414 438]
 [534 567 600]]


### Q2. Z-Score Normalisation

Z-score normalization refers to the process of normalizing every value in a dataset such that the mean of all of the values is 0 and the standard deviation is 1.

We use the following formula to perform a z-score normalization on every value in a dataset:

New value = (x – μ) / σ

where:

x: Original value

μ: Mean of data

σ: Standard deviation of data

##### (i) Without using numpy

In [40]:
def mean(x):
    ans = 0
    for i in range(len(x)):
        ans = ans + x[i]
    ans = ans/len(x)
    return ans

In [50]:
def standard_deviation(x):
  ans1 = 0
  for i in range(len(x)):
    ans1 = ans1 + ((x[i] - mean(x))**2/len(x))
  ans1 = ans1**(0.5)
  return ans1


In [51]:
def zscore_normalisation(x):
  new_matrix1 = [0 for i in x]
  for i in range(len(x)):
    new_matrix1[i]=(x[i]-mean(x))/standard_deviation(x)
  return new_matrix1


##### (ii) Using numpy

Numpy has in_built functions for calculating mean and standard deviation

In [53]:
def zscore_normalisation_2(x):
  mean = np.mean(x)
  std_dev = np.std(x)
  new_matrix2 = [0 for i in x]
  for i in range(len(x)):
    new_matrix2[i]=(x[i]-mean)/std_dev
  return new_matrix2

##### (iii) Testing your code

Run the following cell to check if your functions are working properly.

*Expected Output:* [-1.06753267 -0.99745394 -0.99745394 -0.81057732 -0.41346451 -0.06307086
  0.31068237  0.91803138  1.22170588  1.89913361]

In [56]:
x = [4, 7, 7, 15, 32, 47, 63, 89, 102, 131]
print(f"Result without using numpy: {zscore_normalisation(x)}")
print(f"Result using numpy: {zscore_normalisation_2(x)}")

Result without using numpy: [-1.0675326683028086, -0.9974539373420116, -0.9974539373420116, -0.8105773214465527, -0.4134645126687027, -0.06307085786471742, 0.31068237392620024, 0.9180313755864414, 1.221705876416562, 1.8991336090376003]
Result using numpy: [-1.0675326683028088, -0.9974539373420117, -0.9974539373420117, -0.8105773214465528, -0.41346451266870277, -0.06307085786471743, 0.3106823739262003, 0.9180313755864415, 1.2217058764165623, 1.8991336090376005]


### Q3. Sigmoid fn and its derivative

The sigmoid function is a mathematical function that maps any input value to a value between 0 and 1.

It is defined mathematically as s(x) = 1/(1+e^(-x)).

##### (i) Write a fn to implement sigmoid fn

In [73]:
def sigmoidfn(x):
  e = 2.72
  sgmd_matrix1 = [[0 for p in range(len(x[0]))] for q in range(len(x))]
  for i in range(len(x)):
    for j in range(len(x[0])):
        sgmd_matrix1[i][j] += 1/(1+e**(-x[i][j]))
  return sgmd_matrix1

##### (ii) Write a fn to implement derivative of sigmoid fn

In [81]:
def derivative(x):
    def s(i):
        return 1/(1+e**(-i))
    derivative = np.gradient(s(i),x)
    return derivative
  

##### (iii) Test your code

Run the following cell to check if your functions are working properly.

*Expected output:*

x on applying sigmoid activation fn is: [ [0.99987661 0.88079708 0.99330715 0.5        0.5       ]
 [0.99908895 0.99330715 0.5        0.5        0.5       ] ]

x on applying derivative of sigmoid activation fn is: [ [-1.23379350e-04 -1.04993585e-01 -6.64805667e-03 -2.50000000e-01
  -2.50000000e-01]
 [-9.10221180e-04 -6.64805667e-03 -2.50000000e-01 -2.50000000e-01
  -2.50000000e-01] ]

In [82]:
x = np.array([
    [9,2,5,0,0],
    [7,5,0,0,0]
])
print(f"x on applying sigmoid activation fn is: {sigmoidfn(x)}")
print(f"x on applying derivative of sigmoid activation fn is: {derivative(x)}")

x on applying sigmoid activation fn is: [[0.9998773050820023, 0.8809297008954087, 0.9933281202540091, 0.5, 0.5], [0.9990929659867931, 0.9933281202540091, 0.5, 0.5, 0.5]]


NameError: name 'i' is not defined

# Section 2: Exploring Pandas

*You have been provided with a dataset which includes information about properties of superheated vapor.*

*The dataset consists of the thermophysical properties: specific volume, specific internal energy, specific enthalpy, specific entropy of superheated vapor.*

*Pressure is in kPa and Temperature in centigrade. In the dataframe 75, 100, 125, etc. are temperatures.*

### Read the csv file


In [85]:
df = pd.read_csv('superheated_vapor_properties.csv')


### Display the shape of data frame


In [88]:
df.shape

(544, 37)

### Return an array containing names of all the columns

In [87]:
df.columns

Index(['Pressure', 'Property', 'Liq_Sat', 'Vap_Sat', '75', '100', '125', '150',
       '175', '200', '220', '225', '240', '250', '260', '275', '280', '290',
       '300', '320', '325', '340', '350', '360', '375', '380', '400', '425',
       '450', '475', '500', '525', '550', '575', '600', '625', '650'],
      dtype='object')

### Display the number of null values in each column of the dataframe



In [89]:
df.isnull().sum()

Pressure      0
Property      0
Liq_Sat       0
Vap_Sat       0
75          528
100         508
125         488
150         448
175         384
200         320
220         408
225         400
240         408
250         344
260         384
275         340
280         380
290         488
300          60
320         480
325         136
340         476
350          68
360         476
375         204
380         476
400           0
425         204
450           0
475         204
500           0
525         272
550           0
575         340
600           0
625         476
650           0
dtype: int64

### Create a column which contains the Pressure and Property columns, seperated with 'at' (For eg. V at 1, H at 101.325). Using this print the following:
- Enthalpy at 75 kPa and 573 K
- Entropy at 493 K and 250 kPa



In [ ]:
#your code here

### Find out the column with the highest number of missing values

In [ ]:
#your code here

### What is the average enthalpy of Sat. Liq. at all different pressures in the dataset?

In [ ]:
#your code here

### Separate out the V,U,H,S data from the dataset into V_data, U_data, H_data, S_data

In [ ]:
#your code here

# Section 3: Plotting with Matplotlib

### Plot the properties (specific volume, specific internal energy, specific enthalpy, specific entropy) vs Pressure for saturated liquid.

Note:
- Try using the subplot feature of matplotlib(Explore it!!)
- Provide appropriate title, labels, markersize and other parameters to the plot

In [ ]:
#your code here

### Plot the same for saturated vapor.

In [ ]:
#your code here

### Plot the specific volume of saturated liquid between 300 kPa and 1500 kPa

In [ ]:
#your code here

# Section 4 : Conclusion

*Congratulations on reaching this point! I hope you had fun solving your first assignment and have also built confidence in applying these libraries. If you are wondering, we will cover more about z-score normalization in Week 2, and the sigmoid function will be used in Week 3. After completing this assignment, you are now prepared to learn about machine learning techniques and implement your own machine learning models.*